src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.cs

```diff
@@ -134,12 +134,31 @@ protected void SkipToAndParseCode(Func<HtmlSymbol, bool> condition)
                 {
                     if (last != null)
                     {
-                        Accept(last);
+                        if (!startOfLine || last.Type != HtmlSymbolType.WhiteSpace)
+                        {
+                            Accept(last);
+                        }
+
                         last = null;
                     }
+
                     AddMarkerSymbolIfNecessary();
                     Output(SpanKind.Markup);
+
                     RazorComment();
+
+                    // Handle the whitespace and newline at the end of a razor comment.
+                    if (startOfLine &&
```

Comment:
It'd be worth it to move this bit of logic into `RazorComment` so it can be used/consistent in its other locations (`CSharpCodeParser.cs` and `CSharpCodeParser.Statement.cs`).